File: mobile/apps/photos/plugins/ente_background_manager/ios/Classes/BackgroundManagerPlugin.swift
@preconcurrency import Flutter
import UIKit

@MainActor
public final class BackgroundManagerPlugin: NSObject, @preconcurrency FlutterPlugin {
  private let channel: FlutterMethodChannel
  private var observesOutcomes = false

  private init(registrar: FlutterPluginRegistrar) {
    channel = FlutterMethodChannel(
      name: "io.ente.background/control",
      binaryMessenger: registrar.messenger()
    )
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = BackgroundManagerPlugin(registrar: registrar)
    registrar.addMethodCallDelegate(instance, channel: instance.channel)
    registrar.publish(instance)
    BackgroundRuntime.shared.attach(instance)
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    observesOutcomes = false
    BackgroundRuntime.shared.detach(self)
    channel.setMethodCallHandler(nil)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "configure":
      observesOutcomes = true
      BackgroundRuntime.shared.configure(call.arguments as? [String: Any], result: result)
    case "stopActiveRun":
      BackgroundRuntime.shared.requestStop(result: result)
    case "scheduledTasks":
      BackgroundRuntime.shared.scheduledTasks(result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  var canReport: Bool { observesOutcomes }

  func report(_ event: [String: Any], fallback: @escaping () -> Void) {
    guard observesOutcomes else {
      fallback()
      return
    }
    channel.invokeMethod("outcome", arguments: event) { response in
      if (response as? Bool) != true { fallback() }
    }
  }

  public static func install(
    isEnabled: @escaping () -> Bool,
    registrant: @escaping (FlutterPluginRegistry) -> Void
  ) {
    BackgroundRuntime.shared.install(isEnabled: isEnabled, registrant: registrant)
  }

  @discardableResult
  public static func registerTask(identifier: String, processing: Bool) -> Bool {
    BackgroundRuntime.shared.register(identifier: identifier, processing: processing)
  }
}

File: mobile/apps/photos/plugins/ente_background_manager/android/src/main/kotlin/io/ente/background/BackgroundManagerPlugin.kt
package io.ente.background

import android.app.Application
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class BackgroundManagerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {
    private var channel: MethodChannel? = null
    internal var observesOutcomes = false
        private set

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel =
            MethodChannel(binding.binaryMessenger, CONTROL_CHANNEL).also {
                it.setMethodCallHandler(this)
            }
        BackgroundRuntime.attach(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        observesOutcomes = false
        BackgroundRuntime.detach(this)
        channel?.setMethodCallHandler(null)
        channel = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "configure" -> {
                observesOutcomes = true
                BackgroundRuntime.configure(call.arguments as? Map<*, *>, result)
            }
            "stopActiveRun" -> {
                BackgroundRuntime.requestStop(result)
            }
            "scheduledTasks" -> BackgroundRuntime.scheduledTasks(result)
            else -> result.notImplemented()
        }
    }

    internal fun report(event: Map<String, Any?>, fallback: () -> Unit) {
        val current = channel
        if (!observesOutcomes || current == null) {
            fallback()
            return
        }
        current.invokeMethod(
            "outcome",
            event,
            object : MethodChannel.Result {
                override fun success(result: Any?) {
                    if (result != true) fallback()
                }

                override fun error(code: String, message: String?, details: Any?) = fallback()

                override fun notImplemented() = fallback()
            },
        )
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        BackgroundRuntime.foreground()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        BackgroundRuntime.foreground()
    }

    override fun onDetachedFromActivityForConfigChanges() = Unit

    override fun onDetachedFromActivity() = Unit

    companion object {
        internal const val CONTROL_CHANNEL = "io.ente.background/control"

        fun install(application: Application, isEnabled: () -> Boolean) {
            BackgroundRuntime.install(application, isEnabled)
        }

        fun onForeground() {
            BackgroundRuntime.foreground()
        }
    }
}
